File: ios/EnterpriseMobileApp/Sources/Feature/Rewards/Data/RewardsRepositoryImpl.swift
import Foundation
import CoreDomain
import CoreNetwork

/// Concrete rewards repository with offline caching support.
public final class RewardsRepositoryImpl: RewardsFeatureRepository, @unchecked Sendable {
    private let apiClient: APIClient

    public init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    public func getRewards() -> AsyncStream<DomainResult<[Reward]>> {
        AsyncStream { continuation in
            Task {
                // In production: emit cached rewards from SwiftData first
                do {
                    let response: RewardsResponse = try await apiClient.request(endpoint: "/rewards")
                    let rewards = response.rewards.map { $0.toDomain() }
                    continuation.yield(.success(rewards))
                } catch {
                    continuation.yield(.failure(.serverError(error.localizedDescription)))
                }
                continuation.finish()
            }
        }
    }

    public func getAccount() async -> DomainResult<RewardsAccount> {
        do {
            let response: RewardsAccountDTO = try await apiClient.request(endpoint: "/rewards/account")
            return .success(response.toDomain())
        } catch {
            return .failure(.serverError(error.localizedDescription))
        }
    }

    public func redeemReward(rewardId: String) async -> DomainResult<RewardsAccount> {
        do {
            let response: RewardsAccountDTO = try await apiClient.request(
                endpoint: "/rewards/redeem",
                method: .post,
                body: RedeemRequest(rewardId: rewardId)
            )
            return .success(response.toDomain())
        } catch {
            return .failure(.serverError(error.localizedDescription))
        }
    }

    public func getRewardsByCategory(_ category: RewardCategory) async -> DomainResult<[Reward]> {
        do {
            let response: RewardsResponse = try await apiClient.request(
                endpoint: "/rewards?category=\(category.rawValue)"
            )
            return .success(response.rewards.map { $0.toDomain() })
        } catch {
            return .failure(.serverError(error.localizedDescription))
        }
    }
}

// MARK: - DTOs

struct RewardsResponse: Decodable {
    let rewards: [RewardDTO]
}

struct RewardDTO: Decodable {
    let id: String
    let title: String
    let description: String
    let pointsCost: Int
    let imageUrl: String?
    let expirationDate: String?
    let category: String?

    enum CodingKeys: String, CodingKey {
        case id, title, description, pointsCost, category
        case imageUrl = "image_url"
        case expirationDate = "expiration_date"
    }

    func toDomain() -> Reward {
        Reward(
            id: id,
            title: title,
            description: description,
            pointsCost: pointsCost,
            imageURL: imageUrl.flatMap(URL.init(string:)),
            expirationDate: expirationDate.flatMap { ISO8601DateFormatter().date(from: $0) },
            category: RewardCategory(rawValue: category ?? "general") ?? .general
        )
    }
}

struct RewardsAccountDTO: Decodable {
    let userId: String
    let totalPoints: Int
    let tierLevel: String
    let redeemedRewards: [String]

    func toDomain() -> RewardsAccount {
        RewardsAccount(
            userId: userId,
            totalPoints: totalPoints,
            tierLevel: TierLevel(rawValue: tierLevel) ?? .bronze,
            redeemedRewards: redeemedRewards
        )
    }
}

struct RedeemRequest: Encodable {
    let rewardId: String
}

File: android/feature/rewards/src/main/java/com/enterprise/feature/rewards/data/RewardsRepositoryImpl.kt
package com.enterprise.feature.rewards.data

import com.enterprise.core.domain.entities.DomainError
import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.entities.Reward
import com.enterprise.core.domain.entities.RewardCategory
import com.enterprise.core.domain.entities.RewardsAccount
import com.enterprise.feature.rewards.data.api.RewardsApiService
import com.enterprise.feature.rewards.data.dto.RedeemRequest
import com.enterprise.feature.rewards.domain.repositories.RewardsFeatureRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject
import javax.inject.Singleton

/** Rewards repository with network calls and DTO mapping. */
@Singleton
class RewardsRepositoryImpl @Inject constructor(
    private val rewardsApiService: RewardsApiService
) : RewardsFeatureRepository {

    override fun getRewards(): Flow<DomainResult<List<Reward>>> = flow {
        try {
            val response = rewardsApiService.getRewards()
            val rewards = response.rewards.map { it.toDomain() }
            emit(DomainResult.Success(rewards))
        } catch (e: Exception) {
            emit(DomainResult.Failure(DomainError.NetworkUnavailable))
        }
    }

    override suspend fun getAccount(): DomainResult<RewardsAccount> {
        return try {
            val accountDTO = rewardsApiService.getAccount()
            DomainResult.Success(accountDTO.toDomain())
        } catch (e: Exception) {
            DomainResult.Failure(DomainError.NetworkUnavailable)
        }
    }

    override suspend fun redeemReward(rewardId: String): DomainResult<RewardsAccount> {
        return try {
            val accountDTO = rewardsApiService.redeemReward(RedeemRequest(rewardId))
            DomainResult.Success(accountDTO.toDomain())
        } catch (e: Exception) {
            DomainResult.Failure(DomainError.Unknown(e.localizedMessage ?: "Redeem failed"))
        }
    }

    override suspend fun getRewardsByCategory(category: RewardCategory): DomainResult<List<Reward>> {
        return try {
            val categoryString = category.name.lowercase()
            val response = rewardsApiService.getRewardsByCategory(categoryString)
            DomainResult.Success(response.rewards.map { it.toDomain() })
        } catch (e: Exception) {
            DomainResult.Failure(DomainError.NetworkUnavailable)
        }
    }
}
